### ISSUE TITLE
feature request: could send|receive be made interruptible|resumable?

### ISSUE THREAD

**Issue by calestyo** [kdave/btrfs-progs#720]
Hey.

From [[0](https://lore.kernel.org/linux-btrfs/89811fd8276376cc8b5ef10ec5cec91c36a828f5.camel@scientia.org/)], where it got no attention:

Hey.

This goes along with #719.

Could `send|receive` be extended to allow being interrupted and resumed later on?<br>
Maybe at least when sending read-only snapshots?

I'd naively assume that for a read-only snapshot it's always the same that's being sent (well not really sure, if that's really the case, e.g. when block groups are balanced, defragged, etc.).

If so, `INT`errupting the two could print out some information on how far it already went, which next time one could be feed into `send` as parameter, and it would continue from there.

The motivation for this feature request is, that `send|receving` very large subvolumes takes obviously quite long and one may want/need to pause that (e.g. I don't want to leave my laptop running for days, even while I'm away).

Thanks,
Chris.

[0] https://lore.kernel.org/linux-btrfs/6f340d377c584239a1b8d4236d404d708e61d968.camel@scientia.org/T/#u

**Comment by ChocolateLoverRaj** [comment]
Finding out how many bytes were sent successfully before it was interrupted can be done by having a program in between `send` and the destination.

 The feature that btrfs is lacking is the ability to skip the first `n` bytes when running `send` again. Right now, let's say a total of 100GiB needs to be sent. If it fails after 90GiB, we could just re-run the `send` command and ignore the first 90GiB, but that would cause 90GiB of unnecessary disk reading and would take a very long time. But also it would be nice to have a way of checking that backups were not corrupted, to be certain that files are actually backed up properly.

I am new to ZFS and btrfs (I used btrfs as the `/` partition, but I haven't used it with redundancy, snapshots, backups, etc), and I'm planning on using ZFS or btrfs to save family files locally with redundancy as well as a cloud backup. Incremental backups seem like the best way of doing it, but both ZFS and btrfs lack resuming interrupted `send`s. If anyone knows of an alternative I would appreciate it too.

**Comment by Zygo** [comment]
sends in general aren't that repeatable.  You could run into subtle problems if the extent references in any part of the send changed.  This isn't a problem for `btrfs send -p` because a restarted send will compute a completely new stream that will be consistent with itself.  The problem here is that we're asking for a send stream to be consistent with another send stream that can no longer be recreated.

> I'm planning on using ZFS or btrfs to save family files locally with redundancy as well as a cloud backup...
> If anyone knows of an alternative I would appreciate it too.

Make a ro snapshot on the sender and a rw snapshot on the reeceiver, then use `rsync` for reliably resumable transport to bring the two subvols into sync, then make a ro snapshot of the rw snapshot on the receiver.  `btrfs send` isn't necessarily the best tool for every purpose.

**Comment by ChocolateLoverRaj** [comment]
> Make a ro snapshot on the sender and a rw snapshot on the reeceiver, then use `rsync` for reliably resumable transport to bring the two subvols into sync, then make a ro snapshot of the rw snapshot on the receiver.

I don't think this is possible because I am planning on using AWS S3 Glacier Deep to store the backups.

> `btrfs send` isn't necessarily the best tool for every purpose.

This makes sense. I will look into some programs that build on top of zfs/btrfs for saving backups. Do you have any recommendations for my use case?

**Comment by Forza-tng** [comment]
> > `btrfs send` isn't necessarily the best tool for every purpose.
> 
> This makes sense. I will look into some programs that build on top of zfs/btrfs for saving backups. Do you have any recommendations for my use case?

I wouldn't use cloud storage without encryption. If you want to stay btrfs native, you can store the btrfs send stream as an encrypted file. Btrbk is a tool that can automate this. But there are many backup tools that can utilise btrfs snapshots as source for normal file based backups.

**Comment by calestyo** [comment]
I think we should keep discussions focused on the original request here, and not spread out to general backup discussions.
